### PR TITLE
Prevent StepHandler from redirecting to EmailAuthenticator when user is authenticated

### PR DESCRIPTION
**Pull request by hwupathum**
## Purpose
Prevent StepHandler from redirecting to EmailOTPAuthenticator when the user has authenticated with a different authenticator in the same authentication step.

i.e.

First step -> Identifier First
Second Step -> Basic, Email OTP

Scenario: User signs in with their basic authentication password but are then redirected to Email OTP since the Step Handler invokes the canHandle method which returns true.

## Related PR
- https://github.com/wso2-extensions/identity-outbound-auth-email-otp/pull/156
- https://github.com/wso2/carbon-identity-framework/pull/5055

Related issues
- https://github.com/wso2/product-is/issues/17072
- https://github.com/wso2/product-is/issues/17161

### CODE REVIEW

**Comment by hwupathum** [comment]
**Testing Strategy:**
(1st step -> 2nd step)

- [x] IDF -> Basic / Email
- [x] IDF / Email -> Basic
- [x] Basic ->  Email / SMS
- [x] Email  
- [x] Basic / Email

Using Email Address as the Username 
- [x] IDF -> Basic / Email
- [x] IDF / Email -> Basic
- [x] Basic ->  Email / SMS
- [x] Email  
- [x] Basic / Email

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 1]
```diff
@@ -129,8 +129,18 @@ public boolean canHandle(HttpServletRequest request) {
...
-                || StringUtils.isNotEmpty(request.getParameter(EmailOTPAuthenticatorConstants.EMAIL_ADDRESS))
-                || StringUtils.isNotEmpty(request.getParameter(EmailOTPAuthenticatorConstants.USER_NAME)));
+                || StringUtils.isNotEmpty(request.getParameter(EmailOTPAuthenticatorConstants.EMAIL_ADDRESS)));
+    }
+
+    @Override
+    public boolean canHandleMultiOptionRequest(HttpServletRequest request, AuthenticationContext context) {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Inside EmailOTPAuthenticator canHandleMultiOptionRequest method");
```
Comment (hwupathum) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: Add the debug log with the result

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 2]
```diff
@@ -129,8 +129,18 @@ public boolean canHandle(HttpServletRequest request) {
         return ((StringUtils.isNotEmpty(request.getParameter(EmailOTPAuthenticatorConstants.RESEND))
                 && StringUtils.isEmpty(request.getParameter(EmailOTPAuthenticatorConstants.CODE)))
                 || StringUtils.isNotEmpty(request.getParameter(EmailOTPAuthenticatorConstants.CODE))
-                || StringUtils.isNotEmpty(request.getParameter(EmailOTPAuthenticatorConstants.EMAIL_ADDRESS))
-                || StringUtils.isNotEmpty(request.getParameter(EmailOTPAuthenticatorConstants.USER_NAME)));
+                || StringUtils.isNotEmpty(request.getParameter(EmailOTPAuthenticatorConstants.EMAIL_ADDRESS)));
+    }
+
+    @Override
+    public boolean canHandleMultiOptionRequest(HttpServletRequest request, AuthenticationContext context) {
```
Comment (hwupathum) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: canHandleRequestFromMultiOptionStep

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 3]
```diff
@@ -129,8 +129,18 @@ public boolean canHandle(HttpServletRequest request) {
         return ((StringUtils.isNotEmpty(request.getParameter(EmailOTPAuthenticatorConstants.RESEND))
```
Comment (hwupathum) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java:130]: Add line comments

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 4]
```diff
@@ -129,8 +129,18 @@ public boolean canHandle(HttpServletRequest request) {
...
+
+    @Override
+    public boolean canHandleMultiOptionRequest(HttpServletRequest request, AuthenticationContext context) {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Inside EmailOTPAuthenticator canHandleMultiOptionRequest method");
+        }
+        return canHandle(request) ||
+                (StringUtils.isNotBlank(request.getParameter(EmailOTPAuthenticatorConstants.USER_NAME)) &&
+                        getName().equalsIgnoreCase(context.getCurrentAuthenticator()));
```
Comment (hwupathum) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java:147]: Add inline comment to explain the behaviour

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 5]
```diff
@@ -126,11 +126,31 @@ public boolean canHandle(HttpServletRequest request) {
...
+        // use the currentAuthenticator context to identify if the request is coming from EmailOTP authenticator.
+        // For basic and IDF authenticators, the currentAuthenticator context will be null.
+        boolean canHandleRequest = canHandle(request) ||
+                (StringUtils.isNotBlank(request.getParameter(EmailOTPAuthenticatorConstants.USER_NAME)) &&
+                        getName().equalsIgnoreCase(context.getCurrentAuthenticator()));
+        if (canHandleRequest) {
+            log.debug("EmailOTPAuthenticator can handle the request from multi-option step");
+        } else {
+            log.debug("EmailOTPAuthenticator cannot handle the request from multi-option step");
+        }
```
Comment (pamodaaw) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]:
```suggestion
        if (canHandleRequest) {
            log.debug("EmailOTPAuthenticator can handle the request from multi-option step.");
        } else {
            log.debug("EmailOTPAuthenticator cannot handle the request from multi-option step.");
        }
```

**Review by pamodaaw — APPROVED** [review]

**Pull request merged**